### PR TITLE
Libplugin refactoring

### PR DESCRIPTION
**Pull request by darosior**
In order to correctly (i.e. not loose asynchronicity) implement #3354 I needed libplugin to use `ccan/io`. This escalated quickly to a complete refactoring of `libplugin`.

This PR:
- Moves `json_stream.h` and its corresponding helpers to `common/`, which was required to properly implement `io_plan`s, and which seems generally good as it now allows plugins to use higher-level JSON helpers as in `lightningd/*`.
- Makes `stdin` / `stdout` plugin connections use exclusively `io_plan`s which simplifies the `plugin_main` and keeps the same logic as the other side: `lightningd/plugin`.
- Makes async RPC requests (`send_outreq`) use `io_plan`s, which I believe could help with #3401 though I haven't tested it yet. However I'm happy to revert this patch if that clashes with @ZmnSCPxj work with sparks.
- Overall removes globals in favor of a top-level struct, creatively named `plugin`.

Left for another PR:
- ~~Move existing plugins to use `json_stream` helpers instead of "raw" `jout`s~~ Done in #3480
- Maybe completely get rid of the synchronous RPC methods (`rpc_delve` and co.) ?

Changelog-None

### CODE REVIEW

**Comment by darosior** [comment]
Rebased (and added a1857dfbf1ec8a598d8cda9ad930e1dd42ea209e for errcodes).

**Comment by darosior** [comment]
Re-rebased

File: plugins/libplugin.c [hunk 1]
```diff
@@ -26,21 +24,11 @@
...
 
-/* Map from json command names to usage strings: we don't put this inside
- * struct json_command as it's good practice to have those const. */
-static STRMAP(const char *) usagemap;
-
-/* Timers */
-static struct timers timers;
-static size_t in_timer;
-
-bool deprecated_apis;
```
[the remark this answers is not in the document]
Comment (darosior) [reply plugins/libplugin.c]: But the `plugin` is accessible everywhere ?

File: plugins/libplugin.h [hunk 2]
```diff
@@ -11,17 +12,59 @@
...
 
 enum plugin_restartability {
 	PLUGIN_STATIC,
 	PLUGIN_RESTARTABLE
 };
 
+struct rpc_conn {
+	int fd;
+	MEMBUF(char) mb;
+};
```
[the remark this answers is not in the document]
Comment (darosior) [reply plugins/libplugin.h]: But they dereference it from struct plugin ? At least `pay` and `fundchannel` do, not sure if they do for `rpc_conn` though.

**Comment by darosior** [comment]
I did the change about `rpc_conn` in a new top-commit, if that's ok it avoids me 3 conflicting rebases (this branch, #3480, and the Bitcoin backend plugin's branch).

File: plugins/libplugin.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (rustyrussell) [reply plugins/libplugin.c]: 1. The core code also uses a global, so codesharing is easier this way.
2. Without this, the definition of struct plugin doesn't need to be exposed.

**Comment by rustyrussell** [comment]
Trivial rebase on master for test autogen conflicts...

**Comment by darosior** [comment]
Thanks !

File: plugins/libplugin.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (darosior) [reply plugins/libplugin.c]: Ok, will do tomorrow

**Pull request merged**